File: focus-log/src/lib/sheets/columns.ts
/**
 * Column layout for the focus-log spreadsheet.
 *
 * Every consumer — the CSV template, the A1 ranges used against the Sheets
 * API, the row serialiser and the row parser — is derived from these arrays.
 * That is deliberate: the old code hardcoded "logs!A:C" for writes and
 * "logs!A:D" for reads, so column D (duration) was read but never written
 * (C13). Deriving everything from one list makes that class of drift
 * impossible.
 */

/** A cell as it can arrive from the Sheets API with UNFORMATTED_VALUE. */
export type Cell = string | number | boolean | null | undefined;
export type Row = Cell[];

export type ColumnType = "string" | "number" | "boolean";

export interface ColumnDef<K extends string = string> {
  readonly key: K;
  readonly type: ColumnType;
  /** Documentation for SETUP.md; not used at runtime. */
  readonly note?: string;
}

export const SCHEMA_VERSION = 1;

export const GOAL_COLUMNS = [
  { key: "goal_id", type: "string", note: "UUIDv4. Stable forever, never reused." },
  { key: "title", type: "string" },
  { key: "color", type: "string", note: "Hex colour used by the UI." },
  { key: "weekly_target_minutes", type: "number", note: "0 means no target." },
  { key: "sort_order", type: "number" },
  { key: "status", type: "string", note: "active | archived" },
  { key: "created_at", type: "string", note: "ISO-8601 UTC" },
  { key: "updated_at", type: "string", note: "ISO-8601 UTC. The last-write-wins key." },
  { key: "deleted", type: "boolean", note: "Tombstone. TRUE hides the row." },
  { key: "device_id", type: "string", note: "Tie-breaker for identical updated_at." },
] as const satisfies readonly ColumnDef[];

export const SESSION_COLUMNS = [
  { key: "log_id", type: "string", note: "UUIDv4 minted before the write, so retries are idempotent." },
  { key: "goal_id", type: "string" },
  { key: "start_utc", type: "string", note: "ISO-8601 UTC" },
  { key: "end_utc", type: "string", note: "ISO-8601 UTC" },
  { key: "duration_seconds", type: "number", note: "Computed by the app, not a sheet formula." },
  { key: "local_date", type: "string", note: "YYYY-MM-DD in the timezone at log time." },
  { key: "tz", type: "string", note: "IANA zone, e.g. Asia/Singapore." },
  { key: "note", type: "string" },
  { key: "source", type: "string", note: "timer | manual | import" },
  { key: "updated_at", type: "string", note: "ISO-8601 UTC. The last-write-wins key." },
  { key: "deleted", type: "boolean", note: "Tombstone." },
  { key: "device_id", type: "string", note: "Tie-breaker for identical updated_at." },
] as const satisfies readonly ColumnDef[];

export const META_COLUMNS = [
  { key: "key", type: "string" },
  { key: "value", type: "string" },
] as const satisfies readonly ColumnDef[];

export const TAB_NAMES = {
  goals: "goals",
  sessions: "sessions",
  meta: "meta",
} as const;

export type TabName = (typeof TAB_NAMES)[keyof typeof TAB_NAMES];

/** 0 -> "A", 25 -> "Z", 26 -> "AA". */
export function columnLetter(index: number): string {
  if (index < 0) throw new RangeError(`Column index must be >= 0, got ${index}`);
  let n = index;
  let letters = "";
  for (;;) {
    letters = String.fromCharCode(65 + (n % 26)) + letters;
    if (n < 26) return letters;
    n = Math.floor(n / 26) - 1;
  }
}

/** Whole-column A1 range covering exactly the defined columns, e.g. `sessions!A:L`. */
export function fullRange(tab: TabName, columns: readonly ColumnDef[]): string {
  return `${tab}!A:${columnLetter(columns.length - 1)}`;
}

export function headerRow(columns: readonly ColumnDef[]): string[] {
  return columns.map((c) => c.key);
}

export const RANGES = {
  goals: fullRange(TAB_NAMES.goals, GOAL_COLUMNS),
  sessions: fullRange(TAB_NAMES.sessions, SESSION_COLUMNS),
  meta: fullRange(TAB_NAMES.meta, META_COLUMNS),
} as const;

File: focus-log-wear/core/src/main/kotlin/com/focuslog/core/sheets/Columns.kt
package com.focuslog.core.sheets

/**
 * Column layout for the focus-log spreadsheet — a port of `sheets/columns.ts`.
 *
 * Every consumer (A1 ranges, row serialiser, row parser) derives from these
 * lists. Deriving everything from one source makes read/write column drift
 * impossible — the class of bug that once had column D read but never written.
 *
 * Column order is load-bearing: the app addresses cells by position.
 */

/** A cell as it can arrive from the Sheets API with UNFORMATTED_VALUE. */
typealias Cell = Any?
typealias Row = List<Cell>

enum class ColumnType { STRING, NUMBER, BOOLEAN }

data class ColumnDef(val key: String, val type: ColumnType)

// v2 adds the optional `active` tab (cross-device running timer). Backward
// compatible: a v2 client still reads and writes a v1 sheet, it just leaves the
// cross-device timer off until the `active` tab is added.
const val SCHEMA_VERSION = 2

val GOAL_COLUMNS: List<ColumnDef> = listOf(
    ColumnDef("goal_id", ColumnType.STRING),
    ColumnDef("title", ColumnType.STRING),
    ColumnDef("color", ColumnType.STRING),
    ColumnDef("weekly_target_minutes", ColumnType.NUMBER),
    ColumnDef("sort_order", ColumnType.NUMBER),
    ColumnDef("status", ColumnType.STRING),
    ColumnDef("created_at", ColumnType.STRING),
    ColumnDef("updated_at", ColumnType.STRING),
    ColumnDef("deleted", ColumnType.BOOLEAN),
    ColumnDef("device_id", ColumnType.STRING),
)

val SESSION_COLUMNS: List<ColumnDef> = listOf(
    ColumnDef("log_id", ColumnType.STRING),
    ColumnDef("goal_id", ColumnType.STRING),
    ColumnDef("start_utc", ColumnType.STRING),
    ColumnDef("end_utc", ColumnType.STRING),
    ColumnDef("duration_seconds", ColumnType.NUMBER),
    ColumnDef("local_date", ColumnType.STRING),
    ColumnDef("tz", ColumnType.STRING),
    ColumnDef("note", ColumnType.STRING),
    ColumnDef("source", ColumnType.STRING),
    ColumnDef("updated_at", ColumnType.STRING),
    ColumnDef("deleted", ColumnType.BOOLEAN),
    ColumnDef("device_id", ColumnType.STRING),
)

val META_COLUMNS: List<ColumnDef> = listOf(
    ColumnDef("key", ColumnType.STRING),
    ColumnDef("value", ColumnType.STRING),
)

// The shared running timer (v2+): one versioned row per in-progress session,
// keyed by `log_id`. The id is minted at start and reused at finalise, so two
// devices ending the same timer append rows with the same id that LWW collapses.
// `deleted=TRUE` tombstones a stopped/discarded timer; `segments` carries the
// pause structure so another device shows correct elapsed and can stop it.
val ACTIVE_COLUMNS: List<ColumnDef> = listOf(
    ColumnDef("log_id", ColumnType.STRING),
    ColumnDef("goal_id", ColumnType.STRING),
    ColumnDef("segments", ColumnType.STRING),
    ColumnDef("note", ColumnType.STRING),
    ColumnDef("updated_at", ColumnType.STRING),
    ColumnDef("deleted", ColumnType.BOOLEAN),
    ColumnDef("device_id", ColumnType.STRING),
)

object TabNames {
    const val GOALS = "goals"
    const val SESSIONS = "sessions"
    const val META = "meta"
    const val ACTIVE = "active"
}

/** 0 -> "A", 25 -> "Z", 26 -> "AA". */
fun columnLetter(index: Int): String {
    require(index >= 0) { "Column index must be >= 0, got $index" }
    var n = index
    var letters = ""
    while (true) {
        letters = ('A' + (n % 26)) + letters
        if (n < 26) return letters
        n = n / 26 - 1
    }
}

/** Whole-column A1 range covering exactly the defined columns, e.g. `sessions!A:L`. */
fun fullRange(tab: String, columns: List<ColumnDef>): String =
    "$tab!A:${columnLetter(columns.size - 1)}"

fun headerRow(columns: List<ColumnDef>): List<String> = columns.map { it.key }

object Ranges {
    val goals: String = fullRange(TabNames.GOALS, GOAL_COLUMNS)
    val sessions: String = fullRange(TabNames.SESSIONS, SESSION_COLUMNS)
    val meta: String = fullRange(TabNames.META, META_COLUMNS)
    val active: String = fullRange(TabNames.ACTIVE, ACTIVE_COLUMNS)
}
